android/src/main/java/com/oney/WebRTCModule/DailyWebRTCDevicesManager.java

```diff
@@ -0,0 +1,201 @@
+        }
+        return cameraEnumerator;
+    }
+
+    ReadableArray enumerateDevices() {
+        WritableArray devicesArray = Arguments.createArray();
+        this.fillVideoInputDevices(devicesArray);
+        this.fillAudioDevices(devicesArray);
+        return devicesArray;
+    }
+
+    private void fillVideoInputDevices(WritableArray enumerateDevicesArray){
+        String[] devices = cameraEnumerator.getDeviceNames();
+        for (int i = 0; i < devices.length; ++i) {
+            String deviceName = devices[i];
+            boolean isFrontFacing;
+            try {
+                // This can throw an exception when using the Camera 1 API.
+                isFrontFacing = cameraEnumerator.isFrontFacing(deviceName);
+            } catch (Exception e) {
+                Log.e(TAG, "Failed to check the facing mode of camera");
+                continue;
+            }
+            String label = isFrontFacing ? "Front camera" : "Rear camera";
+            String deviceID = isFrontFacing ? "CAMERA_USER" : "CAMERA_ENVIRONMENT";
+            WritableMap params = this.createWritableMap(deviceID, label, DeviceKind.VIDEO_INPUT.getKind());
+            params.putString("facing", isFrontFacing ? "user" : "environment");
+            enumerateDevicesArray.pushMap(params);
+        }
+    }
+
+    private void fillAudioDevices(WritableArray enumerateDevicesArray){
+        AudioDeviceInfo[] audioOutputDevices = this.audioManager.getDevices(AudioManager.GET_DEVICES_OUTPUTS);
+
+        boolean isWiredHeadsetPlugged = Arrays.stream(audioOutputDevices).anyMatch(
+                device -> device.getType() == AudioDeviceInfo.TYPE_WIRED_HEADSET ||
+                        device.getType() == AudioDeviceInfo.TYPE_WIRED_HEADPHONES
+        );
+
+        WritableMap params = isWiredHeadsetPlugged ?
+                this.createWritableMap(AudioDeviceType.WIRED_OR_EARPIECE.toString(), "Wired headset", DeviceKind.AUDIO.getKind()) :
+                this.createWritableMap(AudioDeviceType.WIRED_OR_EARPIECE.toString(), "Phone Earpiece", DeviceKind.AUDIO.getKind());
+        enumerateDevicesArray.pushMap(params);
+
+        //speaker
+        params = this.createWritableMap(AudioDeviceType.SPEAKER.toString(), "Speaker", DeviceKind.AUDIO.getKind());
+        enumerateDevicesArray.pushMap(params);
+
+        boolean isBluetoothHeadsetPlugged = Arrays.stream(audioOutputDevices).anyMatch(device -> device.getType() == AudioDeviceInfo.TYPE_BLUETOOTH_SCO);
+        if(isBluetoothHeadsetPlugged){
+            params = this.createWritableMap(AudioDeviceType.BLUETOOTH.toString(), "Bluetooth", DeviceKind.AUDIO.getKind());
+            enumerateDevicesArray.pushMap(params);
+        }
+    }
+
+    private WritableMap createWritableMap(String deviceId, String label, String kind){
+        WritableMap audioMap = Arguments.createMap();
+        audioMap.putString("deviceId", deviceId);
+        audioMap.putString("groupId", "");
```

Comment:
if we're deviating from spec and we're not using groupId, should we even bother including it in enumerateDevices?

Reply:
(maybe yes, haven't given it as much thought as you)

Reply:
No strong feelings one way or the other. This is a no-op.